Development/nmos/authorization_behaviour.cpp

```diff
@@ -0,0 +1,523 @@
+                            {
+                                // no token or token expired
+                                auto is_access_token_bad = [&]
+                                {
+                                    auto lock = authorization_state.read_lock();
+
+                                    const auto& bearer_token = authorization_state.bearer_token;
+                                    return (!bearer_token.is_valid_access_token() || is_access_token_expired(bearer_token.access_token(), authorization_state.issuers, authorization_state.authorization_server_uri, gate));
+                                };
+
+                                auto is_client_expired = [&]
+                                {
+                                    // Time at which the client_secret will expire. If time is 0, it will never expire
+                                    // The time is represented as the number of seconds from 1970-01-01T0:0:0Z as measured in UTC
+                                    const auto expires_at = nmos::experimental::fields::client_secret_expires_at(client_metadata);
+                                    if (expires_at == 0)
+                                    {
+                                        return false;
+                                    }
+                                    auto now = std::chrono::system_clock::now();
+                                    auto exp = std::chrono::system_clock::from_time_t(expires_at);
+                                    return (now > exp);
+                                };
+
+                                utility::string_t authorization_flow;
+                                auto validate_openid_client = false;
+                                with_read_lock(model.mutex, [&]
+                                {
+                                    authorization_flow = nmos::experimental::fields::authorization_flow(model.settings);
+                                    validate_openid_client = nmos::experimental::fields::validate_openid_client(model.settings);
+                                });
+
+                                // if using OpenID Connect Authorization server, update the cache client metadata, in case it has been changed (e.g. changed by the system admin)
+                                if (validate_openid_client)
+                                {
+                                    // if OpenID Connect Authorization server is used, client status can be obtained via the Client Configuration Endpoint
+                                    // "The Client Configuration Endpoint is an OAuth 2.0 Protected Resource that MAY be provisioned by the server for a
+                                    //  specific Client to be able to view and update its registered information."
+                                    // see 3.2 of https://openid.net/specs/openid-connect-registration-1_0.html#ClientConfigurationEndpoint
```

Comment:
It's actually section 4 (not 3.2), but maybe better not to mention section number as the link takes you to the relevant section anyway.
```suggestion
                                    // see https://openid.net/specs/openid-connect-registration-1_0.html#ClientConfigurationEndpoint
```